File: array/BinarySearch/boundary/java/BoundarySearch.java
public final class BoundarySearch {
    private BoundarySearch() {}

    public static int lowerBoundRangeHalving(long[] values, long target) {
        int left = 0;
        int right = values.length;

        while (left < right) {
            int mid = left + (right - left) / 2;
            if (values[mid] < target) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }

        return left;
    }

    public static int lowerBoundPowersOfTwo(long[] values, long target) {
        int position = -1;
        int step = 1;
        while (step < values.length) {
            step <<= 1;
        }

        for (; step > 0; step >>= 1) {
            int next = position + step;
            if (next < values.length && values[next] < target) {
                position = next;
            }
        }

        return position + 1;
    }
}

File: array/BinarySearch/boundary/rust/src/boundary_search.rs
pub fn lower_bound_range_halving(values: &[i64], target: i64) -> i32 {
    let mut left: i32 = 0;
    let mut right: i32 = values.len() as i32;

    while left < right {
        let mid = left + (right - left) / 2;
        if values[mid as usize] < target {
            left = mid + 1;
        } else {
            right = mid;
        }
    }

    left
}

pub fn lower_bound_powers_of_two(values: &[i64], target: i64) -> i32 {
    let mut position: i32 = -1;
    let mut step: i32 = 1;
    while step < values.len() as i32 {
        step <<= 1;
    }

    while step > 0 {
        let next = position + step;
        if next < values.len() as i32 && values[next as usize] < target {
            position = next;
        }
        step >>= 1;
    }

    position + 1
}
